src/components/Store.tsx

```diff
@@ -101,54 +135,79 @@ function reducer(state: ChatState, action: A): ChatState {
-                    Object.assign(memo, {
-                        [id]: state.messages[id] || [],
-                    })
-
-                    return memo
-                }, {}),
-                drafts: action.payload.reduce((memo: DraftCollection, { id }: RoomPayload) => {
-                    Object.assign(memo, {
-                        [id]: state.drafts[id] || '',
-                    })
-
-                    return memo
-                }, {}),
+                messages: action.payload.reduce(
+                    (memo: MessagesCollection, { id }: RoomPayload) => {
+                        Object.assign(memo, {
+                            [id]: state.messages[id] || [],
+                        })
+
+                        return memo
+                    },
+                    {}
+                ),
+                drafts: action.payload.reduce(
+                    (memo: DraftCollection, { id }: RoomPayload) => {
+                        Object.assign(memo, {
+                            [id]: state.drafts[id] || '',
+                        })
+
+                        return memo
+                    },
+                    {}
+                ),
             }
         case ActionType.AddRooms:
             return reducer(state, {
                 type: ActionType.SetRooms,
-                payload: [
-                    ...state.rooms,
-                    ...action.payload,
-                ],
+                payload: [...state.rooms, ...action.payload],
             })
         case ActionType.SetMessages:
-            return state.roomId == null ? state : {
+            return state.roomId == null
+                ? state
+                : {
+                      ...state,
+                      messages: {
+                          ...state.messages,
+                          [state.roomId]: action.payload,
+                      },
+                  }
+        case ActionType.AddMessages:
+            return state.roomId == null
+                ? state
+                : reducer(state, {
+                      type: ActionType.SetMessages,
+                      payload: [
+                          ...state.messages[state.roomId],
+                          ...action.payload,
+                      ],
+                  })
+        case ActionType.SetMetamaskAddress:
+            return {
                 ...state,
-                messages: {
-                    ...state.messages,
-                    [state.roomId]: action.payload,
-                },
+                metamaskAddress: action.payload as string,
+            }
+        case ActionType.SetSessionAddress:
+            return {
+                ...state,
+                sessionAddress: action.payload as string,
+            }
+        case ActionType.SetStreamrClient:
+            return {
+                ...state,
+                streamrClient: action.payload as StreamrClient,
```

Comment:
Same here, the `as` part being a little limiting and confusing. What if I want `streamrClient` to be `undefined`?